File: packages/transform-io/typescript/src/read-transform.ts
import {
  BinaryFile,
  TransformList,
  getFileExtension,
  WorkerPoolFunctionResult,
  WorkerPoolFunctionOption,
} from "itk-wasm";

import mimeToTransformIo from "./mime-to-transform-io.js";
import extensionToTransformIo from "./extension-to-transform-io.js";
import transformIoIndex from "./transform-io-index.js";

import ReadTransformOptions from "./read-transform-options.js";
import ReadTransformResult from "./read-transform-result.js";

interface ReaderResult extends WorkerPoolFunctionResult {
  couldRead: boolean;
  transform: TransformList;
}
interface ReaderOptions extends WorkerPoolFunctionOption {
  /** Use float for the parameters value type. The default is double. */
  floatParameters?: boolean;
}
type Reader = (
  serializedTransform: File | BinaryFile,
  options: ReaderOptions
) => Promise<ReaderResult>;

/**
 * Read a transform file format and convert it to the ITK-Wasm file format
 *
 * @param {webWorker} null | webWorker - Web worker to run the pipeline or null to run it in a new worker
 * @param {File | BinaryFile} serializedTransform - Input transform serialized in the file format
 * @param {ReadTransformOptions} options - options to cast the resulting transform type or to only read transform metadata
 *
 * @returns {Promise<ReadTransformResult>} - result object with the transform and the web worker used
 */
async function readTransform(
  serializedTransform: File | BinaryFile,
  options: ReadTransformOptions = {}
): Promise<ReadTransformResult> {
  const mimeType = (serializedTransform as File).type ?? "";
  const fileName =
    (serializedTransform as File).name ??
    (serializedTransform as BinaryFile).path ??
    "fileName";
  const extension = getFileExtension(fileName).toLowerCase();
  let usedWebWorker = options?.webWorker;

  let serializedTransformFile = serializedTransform as BinaryFile;
  if (serializedTransform instanceof Blob) {
    const serializedTransformBuffer = await serializedTransform.arrayBuffer();
    serializedTransformFile = {
      path: serializedTransform.name,
      data: new Uint8Array(serializedTransformBuffer),
    };
  }

  let io = null;
  if (mimeType && mimeToTransformIo.has(mimeType)) {
    io = mimeToTransformIo.get(mimeType);
  } else if (extensionToTransformIo.has(extension)) {
    io = extensionToTransformIo.get(extension);
  } else {
    for (const readerWriter of transformIoIndex.values()) {
      if (readerWriter[0] !== null) {
        let {
          webWorker: testWebWorker,
          couldRead,
          transform,
        } = await (readerWriter[0] as unknown as Reader)(
          {
            path: serializedTransformFile.path,
            data: serializedTransformFile.data.slice(),
          },
          {
            floatParameters: options.floatParameters,
            webWorker: usedWebWorker,
            noCopy: options?.noCopy,
          }
        );
        usedWebWorker = testWebWorker;
        if (couldRead) {
          return { webWorker: usedWebWorker, transform };
        }
      }
    }
  }
  if (!io) {
    throw Error("Could not find IO for: " + fileName);
  }
  const readerWriter = transformIoIndex.get(io as string);

  const reader = (readerWriter as Array<Reader>)[0];
  let {
    webWorker: testWebWorker,
    couldRead,
    transform,
  } = await reader(serializedTransformFile, {
    floatParameters: options.floatParameters,
    webWorker: usedWebWorker,
    noCopy: options?.noCopy,
  });
  usedWebWorker = testWebWorker;
  if (!couldRead) {
    throw Error("Could not read: " + fileName);
  }

  return { webWorker: usedWebWorker, transform };
}

export default readTransform;

File: packages/transform-io/python/itkwasm-transform-io-wasi/itkwasm_transform_io_wasi/read_transform.py
import os
import importlib
from pathlib import Path

from itkwasm import TransformList

from .extension_to_transform_io import extension_to_transform_io
from .transform_io_index import transform_io_index

def read_transform(
    serialized_transform: os.PathLike,
    float_parameters: bool = False,
) -> TransformList:
    """Read a transform file format and convert it to the ITK-Wasm file format.

    :param serialized_transform: Input transform serialized in the file format
    :type  serialized_transform: os.PathLike

    :param float_parameters: Use float for the parameter value type. The default is double.
    :type  float_parameters: bool

    :return: Output transform
    :rtype:  TransformList
    """
    extension = ''.join(Path(serialized_transform).suffixes)

    io = None
    if extension in extension_to_transform_io:
        func = f"{extension_to_transform_io[extension]}_read_transform"
        mod_name = f"itkwasm_transform_io_wasi.{func}"
        mod = importlib.import_module(mod_name)
        io = getattr(mod, func)
    else:
        for ioname in transform_io_index:
            func = f"{ioname}_read_transform"
            mod_name = f"itkwasm_transform_io_wasi.{func}"
            mod = importlib.import_module(mod_name)
            io = getattr(mod, func)
            could_read, transform = io(serialized_transform, float_parameters=float_parameters)
            if could_read:
                return transform

    if io is None:
        raise RuntimeError(f"Could not find an transform reader for {extension}")

    could_read, transform = io(serialized_transform, float_parameters=float_parameters)
    if not could_read:
        raise RuntimeError(f"Could not read {serialized_transform}")

    return transform


def transformread(
    serialized_transform: os.PathLike,
    float_parameters: bool = False,
) -> TransformList:
    return read_transform(serialized_transform, float_parameters=float_parameters)

transformread.__doc__ = f"""{read_transform.__doc__}
    Alias for read_transform.
    """